### PR TITLE
Fix some overflows and windows warnings

### PR DESCRIPTION
**Pull request by upsj**
When I looked at a few of the warnings that the MSVC compiler generates, I realized that many of them point to potential overflows with 64bit indices, so I will use this PR to collect fixes for these issues.

### CODE REVIEW

**Review by pratikvn — APPROVED** [review]
Nice work on this! LGTM! Is this WIP because you still need to update some more files ?

**Comment by upsj** [comment]
Yes, this needs a lot more work, I only picked a few low-hanging fruits

**Review by yhmtsai — APPROVED** [review]
LGTM!
I think changing float->double in hybrid should be fine because we only use cpu(host) to split the matrix. If we use it in the kernel, need to use float/double explicitly to avoid the protential simulation issue in DPCPP.

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 1]
```diff
@@ -121,8 +121,8 @@ class ResidualNormReduction : public ResidualNorm<ValueType> {
         /**
          * Factor by which the residual norm will be reduced
          */
-        remove_complex<ValueType> GKO_FACTORY_PARAMETER_SCALAR(reduction_factor,
-                                                               1e-15);
+        remove_complex<ValueType> GKO_FACTORY_PARAMETER_SCALAR(
+            reduction_factor, static_cast<remove_complex<ValueType>>(1e-15));
```
Comment (tcojean) [inline include/ginkgo/core/stop/residual_norm.hpp:125]: It's only defined in `core/test/utils.hpp` right now, but should we use the `reduction_factor` from there for the default value here?

**Review by tcojean — APPROVED** [review]
LGTM, thanks for fixing this.

Some time ago I looked at a Windows trace and it's unreadable due to all the warnings, most of which I think come down to a few things.

File: include/ginkgo/core/matrix/hybrid.hpp [hunk 2]
```diff
@@ -276,7 +276,7 @@ class Hybrid
         auto get_percentage() const { return percent_; }
 
     private:
-        float percent_;
+        double percent_;
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/hybrid.hpp:279]: IIRC it complains in some tests when we pass in a double value.

File: reference/solver/upper_trs_kernels.cpp [hunk 3]
```diff
@@ -100,10 +100,12 @@ void solve(std::shared_ptr<const ReferenceExecutor> exec,
     auto col_idxs = matrix->get_const_col_idxs();
     auto vals = matrix->get_const_values();
 
-    for (int j = 0; j < b->get_size()[1]; ++j) {
-        for (int row = matrix->get_size()[0] - 1; row >= 0; --row) {
+    for (size_type j = 0; j < b->get_size()[1]; ++j) {
+        for (size_type inv_row = 0; inv_row < matrix->get_size()[0];
+             ++inv_row) {
+            auto row = matrix->get_size()[0] - 1 - inv_row;
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/solver/upper_trs_kernels.cpp:106]: This was a combination of a signed-unsigned warning and the fact that unsigned >= 0 is always true, so I needed to rewrite the condition to count upwards.

File: include/ginkgo/core/stop/residual_norm.hpp (hunk 1 above)
Comment (upsj) [reply include/ginkgo/core/stop/residual_norm.hpp:125]: Yes, but I would prefer to do that in another PR

**Pull request merged**